test/core/credentials/transport/tls/spiffe_utils_test.cc

```diff
@@ -0,0 +1,281 @@
+//
+//
+// Copyright 2025 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//
+
+#include "src/core/credentials/transport/tls/spiffe_utils.h"
+
+#include <grpc/grpc.h>
+
+#include <string>
+
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_split.h"
+#include "absl/strings/string_view.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "test/core/test_util/test_config.h"
+
+namespace grpc_core {
+namespace testing {
+
+TEST(SpiffeId, EmptyFails) {
+  EXPECT_EQ(
+      SpiffeId::FromString("").status(),
+      absl::InvalidArgumentError("SPIFFE ID cannot be parsed from empty URI"));
+}
+
+TEST(SpiffeId, TooLongFails) {
+  EXPECT_EQ(SpiffeId::FromString(std::string(2049, 'a')).status(),
+            absl::InvalidArgumentError(
+                "URI length is 2049, maximum allowed for SPIFFE ID is 2048"));
+}
+
+TEST(SpiffeId, ContainsHashtagFails) {
+  EXPECT_EQ(
+      SpiffeId::FromString("ab#de").status(),
+      absl::InvalidArgumentError("SPIFFE ID cannot contain query fragments"));
+}
+
+TEST(SpiffeId, ContainsQuestionMarkFails) {
+  EXPECT_EQ(
+      SpiffeId::FromString("ab?de").status(),
+      absl::InvalidArgumentError("SPIFFE ID cannot contain query parameters"));
+}
+
+TEST(SpiffeId, DoesNotStartWithSpiffeFails) {
+  EXPECT_EQ(SpiffeId::FromString("www://foo/bar").status(),
+            absl::InvalidArgumentError("SPIFFE ID must start with spiffe://"));
+}
+
+TEST(SpiffeId, EndWithSlashFails) {
+  EXPECT_EQ(SpiffeId::FromString("spiffe://foo/bar/").status(),
+            absl::InvalidArgumentError("SPIFFE ID cannot end with a /"));
+}
+
+TEST(SpiffeId, NoTrustDomainFails) {
+  EXPECT_EQ(SpiffeId::FromString("spiffe://").status(),
+            absl::InvalidArgumentError("SPIFFE ID cannot end with a /"));
+}
+
+TEST(SpiffeId, NoTrustDomainWithPathFails) {
+  EXPECT_EQ(SpiffeId::FromString("spiffe:///path").status(),
+            absl::InvalidArgumentError("The trust domain cannot be empty"));
+}
+
+TEST(SpiffeId, TrustDomainTooLongFails) {
+  EXPECT_EQ(
+      SpiffeId::FromString(absl::StrCat("spiffe://", std::string(256, 'a')))
+          .status(),
+      absl::InvalidArgumentError(
+          "Trust domain maximum length is 255 characters"));
+}
+
+TEST(SpiffeId, TrustDomainInvalidCharacterFails) {
```

Comment:
Depending on the implementation, this could be subtle, because `@` usually denotes the start of the user info of a URI's hostname. For that reason, could we (a) rename this test to `TrustDomainWithUserInfoFails`, (b) change the string to `spiffe://domain@userinfo` to make this more clear, and (c) add a different test where it fails due to an invalid character (e.g. using $ instead of @)?

Reply:
Done